src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -111,6 +111,7 @@ protected IDictionary CapNames
         internal bool _refsInitialized = false;
 
         internal static LinkedList<CachedCodeEntry> s_livecode = new LinkedList<CachedCodeEntry>();// the cache of code and factories that are currently loaded
+        internal static Dictionary<CachedCodeEntryKey, CachedCodeEntry> s_livecode_dict = new Dictionary<CachedCodeEntryKey, CachedCodeEntry>();
```

Comment:
You might also look at getting rid of s_livecode altogether and just making the CachedCodeEntry themselves into nodes in a linked list (then storing a head and last field directly).